[Writing Tools] macOS: Selection should be hidden when the pondering animation begins, stay hidden until all animations complete

rr-codes opened the pull request:
#### a42d8d1b580267a1428d6056c315e7cf3853ee50
<pre>
[Writing Tools] macOS: Selection should be hidden when the pondering animation begins, stay hidden until all animations complete
<a href="https://bugs.webkit.org/show_bug.cgi?id=277873">https://bugs.webkit.org/show_bug.cgi?id=277873</a>
<a href="https://rdar.apple.com/130995812">rdar://130995812</a>

Reviewed by Wenson Hsieh.

Clear the selection when a session begins, and reveal it when the session replacement animation is completely done.

To facilitate the latter, the UI process now maintains a counter of ongoing intelligence text animations. There are
several distinct sequence of events that should cause the selection to be revealed, assuming a replacement is done
in multiple pieces;

1) The sequence

a. `didReceive` [finished=false]
b. animation 1 starts
c. animation 1 ends
d. `didReceive` [finished=false]
e. animation 2 starts
f. animation 2 ends
g. `didReceive` [finished=true].

In this case, all animations are complete before `finished` is true. Therefore, by (g), the counter will be 0, and
the call from `didReceive` with `finished = true` can reveal the selection.

2) The sequence

a. `didReceive` [finished=false]
b. `didReceive` [finished=false]
c. `didReceive` [finished=true].
d. animation 1 starts
e. animation 1 ends
f. animation 2 starts
g. animation 2 ends

In this case, the call from `didReceive` where `finished = true` happens prior to all the animations finishing. In this
case, the UI process notes that the replacement is finished. Then, in the text animation manager, when the last animation
finishes (which is known via maintaining the counter of animations), then `finished` will be true, and the selection will
be revealed.

Drive-by fix: Fix and improve some formatting across various files.

* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::addSourceTextAnimation):
(WebCore::ChromeClient::addDestinationTextAnimation):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::showSelectionForWritingToolsSessionWithID const):
* Source/WebCore/page/Page.h:
* Source/WebCore/page/writing-tools/WritingToolsController.h:
* Source/WebCore/page/writing-tools/WritingToolsController.mm:
(WebCore::WritingToolsController::willBeginWritingToolsSession):
(WebCore::WritingToolsController::showSelectionForWritingToolsSessionWithID const):
(WebCore::WritingToolsController::compositionSessionDidFinishReplacement):
(WebCore::WritingToolsController::compositionSessionDidReceiveTextWithReplacementRangeAsync):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _addTextAnimationForAnimationID:withData:]):
(-[WKWebView _removeTextAnimationForAnimationID:]):
(-[WKWebView _deferIntelligenceTextAnimation]):
(-[WKWebView willBeginWritingToolsSession:requestContexts:]):
(-[WKWebView compositionSession:didReceiveText:replacementRange:inContext:finished:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::deferIntelligenceTextAnimation):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::enableSourceTextAnimationAfterElementWithID):
(WebKit::WebPageProxy::getTextIndicatorForID):
(WebKit::WebPageProxy::updateUnderlyingTextVisibilityForTextAnimationID):
(WebKit::WebPageProxy::showSelectionForWritingToolsSessionAssociatedWithAnimationID):
(WebKit::WebPageProxy::showSelectionForWritingToolsSessionWithID):
(WebKit::WebPageProxy::deferIntelligenceTextAnimation):
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm:
(-[WKTextAnimationManager addTextAnimationForAnimationID:withData:]):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::hasContentRelativeChildViews const):
(WebKit::WebViewImpl::suppressContentRelativeChildViews):
(WebKit::WebViewImpl::restoreContentRelativeChildViews):
(WebKit::WebViewImpl::addTextAnimationForAnimationID):
(WebKit::WebViewImpl::removeTextAnimationForAnimationID):
(WebKit::WebViewImpl::writingToolsSessionWillBegin):
(WebKit::WebViewImpl::writingToolsCompositionSessionDidReceiveReplacements):
(WebKit::WebViewImpl::isWritingToolsTextReplacementsFinished const):
(WebKit::WebViewImpl::isIntelligenceTextPonderingAnimationFinished const):
(WebKit::WebViewImpl::willBeginPartialIntelligenceTextPonderingAnimation):
(WebKit::WebViewImpl::didEndPartialIntelligenceTextPonderingAnimation):
* Source/WebKit/WebKitSwift/TextAnimation/TextAnimationManager.swift:
(TextAnimationManager.targetedPreview(for:)):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::addSourceTextAnimation):
(WebKit::WebChromeClient::addDestinationTextAnimation):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm:
(WebKit::remainingCharacterRange):
(WebKit::TextAnimationController::addSourceTextAnimation):
(WebKit::TextAnimationController::addDestinationTextAnimation):
(WebKit::TextAnimationController::showSelectionForWritingToolsSessionAssociatedWithAnimationID):
(WebKit::TextAnimationController::updateUnderlyingTextVisibilityForTextAnimationID):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::addTextAnimationForAnimationID):
(WebKit::WebPage::removeTextAnimationForAnimationID):
(WebKit::WebPage::removeTransparentMarkersForSessionID):
(WebKit::WebPage::removeInitialTextAnimation):
(WebKit::WebPage::addInitialTextAnimation):
(WebKit::WebPage::addSourceTextAnimation):
(WebKit::WebPage::addDestinationTextAnimation):
(WebKit::WebPage::clearAnimationsForSessionID):
(WebKit::WebPage::showSelectionForWritingToolsSessionAssociatedWithAnimationID):
(WebKit::WebPage::showSelectionForWritingToolsSessionWithID):
(WebKit::WebPage::deferIntelligenceTextAnimation):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::addTextAnimationForAnimationID): Deleted.
(WebKit::WebPage::removeTextAnimationForAnimationID): Deleted.
(WebKit::WebPage::removeTransparentMarkersForSessionID): Deleted.
(WebKit::WebPage::removeInitialTextAnimation): Deleted.
(WebKit::WebPage::addInitialTextAnimation): Deleted.
(WebKit::WebPage::addSourceTextAnimation): Deleted.
(WebKit::WebPage::addDestinationTextAnimation): Deleted.
(WebKit::WebPage::clearAnimationsForSessionID): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/282125@main">https://commits.webkit.org/282125@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/39edb8342b73bc1bde447285f273621be866010c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62094 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/41448 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14686 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/66074 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/12639 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/64213 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/49134 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12979 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/50008 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8733 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/65163 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38469 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53775 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30839 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/35113 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/11011 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/11570 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56908 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/11315 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67802 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/6037 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/11079 "Found 2 new test failures: imported/w3c/web-platform-tests/css/css-view-transitions/navigation/hide-before-reveal.html workers/worker-to-worker.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/57385 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/6063 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53724 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57635 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4949 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9354 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/37248 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/38332 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/39428 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/38077 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/page/writing-tools/WritingToolsController.mm:
```diff
@@ -355,35 +355,80 @@
+        return;
+    }
+
+    auto* state = std::get_if<CompositionState>(&it->value);
```
Nit - would a `CheckedPtr` be appropriate here? (Maybe with a scoped lifetime so that it's destroyed before the call to `setSelection`, which might update layout and destroy the composition state)

rr-codes replied:
the CompositionState lasts the lifetime of the writing tools controller, since it's just a simple struct type member variable of the controller. This is also in dozens of other places across the file, so I will see if they can all be CheckPtr'd and if so change them all at once in a follow up

whsieh replied:
Sounds good. We should probably deploy smart pointers here to enforce this invariant:

> CompositionState lasts the lifetime of the writing tools controller

whsieh commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -4644,18 +4644,56 @@ static NSPasteboardName pasteboardNameForAccessCategory(WebCore::DOMPasteAccessC
+        return;
+    }
+
+    m_partialIntelligenceTextPonderingAnimationCount -= 1;
```
Should we assert that this never dips below 0?

rr-codes replied:
isn't that what the assertion right above does?

whsieh replied:
Ah, I totally missed that 😓. My bad!

rr-codes committed:
[Writing Tools] macOS: Selection should be hidden when the pondering animation begins, stay hidden until all animations complete
https://bugs.webkit.org/show_bug.cgi?id=277873
rdar://130995812

Reviewed by Wenson Hsieh.

Clear the selection when a session begins, and reveal it when the session replacement animation is completely done.

To facilitate the latter, the UI process now maintains a counter of ongoing intelligence text animations. There are
several distinct sequence of events that should cause the selection to be revealed, assuming a replacement is done
in multiple pieces;

1) The sequence

a. `didReceive` [finished=false]
b. animation 1 starts
c. animation 1 ends
d. `didReceive` [finished=false]
e. animation 2 starts
f. animation 2 ends
g. `didReceive` [finished=true].

In this case, all animations are complete before `finished` is true. Therefore, by (g), the counter will be 0, and
the call from `didReceive` with `finished = true` can reveal the selection.

2) The sequence

a. `didReceive` [finished=false]
b. `didReceive` [finished=false]
c. `didReceive` [finished=true].
d. animation 1 starts
e. animation 1 ends
f. animation 2 starts
g. animation 2 ends

In this case, the call from `didReceive` where `finished = true` happens prior to all the animations finishing. In this
case, the UI process notes that the replacement is finished. Then, in the text animation manager, when the last animation
finishes (which is known via maintaining the counter of animations), then `finished` will be true, and the selection will
be revealed.

Drive-by fix: Fix and improve some formatting across various files.

* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::addSourceTextAnimation):
(WebCore::ChromeClient::addDestinationTextAnimation):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::showSelectionForWritingToolsSessionWithID const):
* Source/WebCore/page/Page.h:
* Source/WebCore/page/writing-tools/WritingToolsController.h:
* Source/WebCore/page/writing-tools/WritingToolsController.mm:
(WebCore::WritingToolsController::willBeginWritingToolsSession):
(WebCore::WritingToolsController::showSelectionForWritingToolsSessionWithID const):
(WebCore::WritingToolsController::compositionSessionDidFinishReplacement):
(WebCore::WritingToolsController::compositionSessionDidReceiveTextWithReplacementRangeAsync):
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _addTextAnimationForAnimationID:withData:]):
(-[WKWebView _removeTextAnimationForAnimationID:]):
(-[WKWebView _deferIntelligenceTextAnimation]):
(-[WKWebView willBeginWritingToolsSession:requestContexts:]):
(-[WKWebView compositionSession:didReceiveText:replacementRange:inContext:finished:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewInternal.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.h:
* Source/WebKit/UIProcess/Cocoa/PageClientImplCocoa.mm:
(WebKit::PageClientImplCocoa::deferIntelligenceTextAnimation):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::enableSourceTextAnimationAfterElementWithID):
(WebKit::WebPageProxy::getTextIndicatorForID):
(WebKit::WebPageProxy::updateUnderlyingTextVisibilityForTextAnimationID):
(WebKit::WebPageProxy::showSelectionForWritingToolsSessionAssociatedWithAnimationID):
(WebKit::WebPageProxy::showSelectionForWritingToolsSessionWithID):
(WebKit::WebPageProxy::deferIntelligenceTextAnimation):
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm:
(-[WKTextAnimationManager addTextAnimationForAnimationID:withData:]):
* Source/WebKit/UIProcess/mac/WebViewImpl.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::hasContentRelativeChildViews const):
(WebKit::WebViewImpl::suppressContentRelativeChildViews):
(WebKit::WebViewImpl::restoreContentRelativeChildViews):
(WebKit::WebViewImpl::addTextAnimationForAnimationID):
(WebKit::WebViewImpl::removeTextAnimationForAnimationID):
(WebKit::WebViewImpl::writingToolsSessionWillBegin):
(WebKit::WebViewImpl::writingToolsCompositionSessionDidReceiveReplacements):
(WebKit::WebViewImpl::isWritingToolsTextReplacementsFinished const):
(WebKit::WebViewImpl::isIntelligenceTextPonderingAnimationFinished const):
(WebKit::WebViewImpl::willBeginPartialIntelligenceTextPonderingAnimation):
(WebKit::WebViewImpl::didEndPartialIntelligenceTextPonderingAnimation):
* Source/WebKit/WebKitSwift/TextAnimation/TextAnimationManager.swift:
(TextAnimationManager.targetedPreview(for:)):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::addSourceTextAnimation):
(WebKit::WebChromeClient::addDestinationTextAnimation):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm:
(WebKit::remainingCharacterRange):
(WebKit::TextAnimationController::addSourceTextAnimation):
(WebKit::TextAnimationController::addDestinationTextAnimation):
(WebKit::TextAnimationController::showSelectionForWritingToolsSessionAssociatedWithAnimationID):
(WebKit::TextAnimationController::updateUnderlyingTextVisibilityForTextAnimationID):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::addTextAnimationForAnimationID):
(WebKit::WebPage::removeTextAnimationForAnimationID):
(WebKit::WebPage::removeTransparentMarkersForSessionID):
(WebKit::WebPage::removeInitialTextAnimation):
(WebKit::WebPage::addInitialTextAnimation):
(WebKit::WebPage::addSourceTextAnimation):
(WebKit::WebPage::addDestinationTextAnimation):
(WebKit::WebPage::clearAnimationsForSessionID):
(WebKit::WebPage::showSelectionForWritingToolsSessionAssociatedWithAnimationID):
(WebKit::WebPage::showSelectionForWritingToolsSessionWithID):
(WebKit::WebPage::deferIntelligenceTextAnimation):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::addTextAnimationForAnimationID): Deleted.
(WebKit::WebPage::removeTextAnimationForAnimationID): Deleted.
(WebKit::WebPage::removeTransparentMarkersForSessionID): Deleted.
(WebKit::WebPage::removeInitialTextAnimation): Deleted.
(WebKit::WebPage::addInitialTextAnimation): Deleted.
(WebKit::WebPage::addSourceTextAnimation): Deleted.
(WebKit::WebPage::addDestinationTextAnimation): Deleted.
(WebKit::WebPage::clearAnimationsForSessionID): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: https://commits.webkit.org/282125@main
